File: leetcode/c++/20.cpp
// Hi, I'm Yanzhan. For more algothmic problems, visit my Youtube Channel (Yanzhan Yang's Youtube Channel) : https://www.youtube.com/channel/UCDkz-__gl3frqLexukpG0DA?view_as=subscriber or my Twitter Account (Yanzhan Yang's Twitter) : https://twitter.com/YangYanzhan or my GitHub HomePage (Yanzhan Yang's GitHub HomePage) : https://yangyanzhan.github.io .
// For this specific algothmic problem, visit my Youtube Video :  .
// It's fascinating to solve algothmic problems, follow Yanzhan to learn more!

class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        for (char ch : s) {
            if (ch == '(') {
                stack.push_back(ch);
            } else if (ch == '[') {
                stack.push_back(ch);
            } else if (ch == '{') {
                stack.push_back(ch);
            } else if (ch == ')') {
                if (!stack.empty() && stack.back() == '(') {
                    stack.pop_back();
                } else {
                    return false;
                }
            } else if (ch == ']') {
                if (!stack.empty() && stack.back() == '[') {
                    stack.pop_back();
                } else {
                    return false;
                }
            } else if (ch == '}') {
                if (!stack.empty() && stack.back() == '{') {
                    stack.pop_back();
                } else {
                    return false;
                }
            }
        }
        return stack.empty();
    }
};

File: leetcode/python/archive/20.py
#!/usr/bin/env python

__author__ = "yangyanzhan"
__email__ = "yangyanzhan@gmail.com"
__url__ = "https://github.com/yangyanzhan/code-camp"
__blog__ = "https://yangyanzhan.github.io"
__youtube__ = "https://www.youtube.com/channel/UCDkz-__gl3frqLexukpG0DA?view_as=subscriber"
__twitter__ = "https://twitter.com/YangYanzhan"

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        l_tokens = "([{"
        r_tokens = ")]}"
        for ch in s:
            idx = l_tokens.find(ch)
            if idx >= 0:
                stack.append(idx)
            else:
                idx = r_tokens.find(ch)
                if len(stack) >= 1 and stack[-1] == idx:
                    del stack[-1]
                else:
                    return False
        return len(stack) == 0
